### ISSUE TITLE
cpp-client loses spans on short-lived processes; Close() not flushing buffers?

### ISSUE THREAD

**Issue by ringerc** [jaegertracing/cpp-client#52]
If a process:

* Creates a tracer with a typical configuration (like the following yml)
* Creates a span
* Finishes the span
* Closes the tracer
* Exits

the span should be reported to the collector. But it isn't unless a short sleep is inserted *before* the `tracer->Close()` call.

If the `sleep()` is added *after*  the tracer close, the spans are not sent. It looks like Close() must be failing to flush buffers, instead dropping any pending buffers on the floor.

Configuration, loaded with the YAML config support:

```
disabled: false
sampler:
    type: const
    param: 1
reporter:
    queueSize: 100
    bufferFlushInterval: 10
    logSpans: false
    localAgentHostPort: 127.0.0.1:6831
headers:
    jaegerDebugHeader: debug-id
    jaegerBaggageHeader: baggage
    TraceContextHeaderName: trace-id
    traceBaggageHeaderPrefix: "testctx-"
baggage_restrictions:
    denyBaggageOnInitializationFailure: false
    hostPort: 127.0.0.1:5778
    refreshInterval: 60
```

**Comment by ringerc** [comment]
Related #53

**Comment by ringerc** [comment]
`RemoteReport::close()` makes no effort to flush by the looks:

```
void RemoteReporter::close()
{
    {
        std::unique_lock<std::mutex> lock(_mutex);
        if (!_running) {
            return;
        }
        _running = false;
        lock.unlock();
        _cv.notify_one();
    }
    _thread.join();
}
```

**Comment by ringerc** [comment]
It probably makes sense to log dropped buffers on close instead, with a hint that an explicit flush before close is advised (using #53 for flush). But that'd really want the opentracing API to adopt the Flush method.

**Comment by rnburn** [comment]
The `Close` method is meant to be optional; so if you don't want to wait on the tracer to flush all it's spans you can just not call it.

I think the right solution is for jaeger to move the current code in `close` to its destructor and change `close` to flush the spans, but require it to be called explicitly.

**Comment by ringerc** [comment]
@rnburn Makes sense. I'll amend my push that adds an optional `Tracer.flush()` to Jaeger's tracer so that it also calls it on explicit `Close()`, but from the dtor.

@isaachier I've written a small test program to reproduce this bug. It reliably loses spans. https://github.com/ringerc/jaeger-cpp-client/commit/2a36dd57864f612454b1bbd0f8c6e9b4fc728190 (on master) or https://github.com/ringerc/jaeger-cpp-client/commit/ff654d91c8b2088f29b910be50fe721180ec550e (on my fix/feature integration branch, but not using flush).

`jaegertracing::Tracer::flush()` does not seem to be working/sufficient.  If you uncomment

```
        //static_cast<jaegertracing::Tracer*>(tracer.get())->flush();
```

in `src/jaegertracing/shortlived.cpp`, it still often loses spans. Wireshark will show that the span was never sent.

**Comment by ringerc** [comment]
I pushed an improved test case. It clearly shows that we lose spans unless:

- We explicitly `flush()` with `./src/jaegertracing/shortlived "" 1`
- We sleep > 10 seconds with `./src/jaegertracing/shortlived "" 0 11`

Also, the flush must take effect. Sometimes it seems to get missed;

```
Test with opname "shortlived--flush-sleep0s, flush 1
flushing...  flushed (flush took 0ms)
not sleeping before exit 
```

in which case the span is also lost. So there's something wrong with my flush method. Locking issues?

@isaachier I could use advice at this point, my C++ and threading is ... limited

**Comment by isaachier** [comment]
In general there might be a problem with blocking IO and threading. I'm about to call it a night, but will try to look at this tomorrow. Worst case, will be able to look over the weekend.

**Comment by isaachier** [comment]
I think I have a simpler approach for this specific issue. About the flush method being exposed overall, that can be a new PR. See #59 for first fix.
